### DISCUSSION TITLE
workspace.applyEdit causing test timeouts

### DISCUSSION THREAD

**Discussion by kimadeline** [microsoft/vscode-discussions#326] [Extension Development QnA]
Hi everyone!

I have a test that recently (end of November, beginning of December) started consistently causing timeouts when run in GitHub actions (`Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.` kind of thing). 

The "fun" part is that the test in itself doesn't time out; rather, it's the next test suite that times out in its `beforeEach`/`setup phase` 🥲 We attach a listener to `onDidChangeTextDocument` as part of the class being tested, but even when I comment it out the timeout still happens. If I rerun the test the timeout sometimes disappear, but not always.

The test (without any extension-specific code) looks like this:

```ts
test('It does a thing', async () => {
        const content = dedent`
            function foo() {
                return 1;
            }`;

        const document = await workspace.openTextDocument({
            language: 'typescript',
            content,
        });

        const edit = new WorkspaceEdit();
        edit.insert(document.uri, new Position(0, 0), 'abcd');
        await workspace.applyEdit(edit);

        await new Promise(resolve => setTimeout(resolve, 100));
    });
```

This happens with VS Code 1.73.1, 1.74.1, 1.74.2 and 1.75-insiders, mainly on Ubuntu 22.04 (ubuntu-latest), and a bit less frequently on Windows (windows-latest).

**Comment by Drisate** [comment]
There are a few potential causes for this type of behavior:

The test itself might be taking too long to run, which could cause a timeout. You could try adding some debugging statements to see how long the test takes to run, and compare that to the timeout threshold.

There might be an issue with the test environment, such as resource contention or a race condition. In this case, you could try running the test in a different environment to see if the behavior persists.

There might be an issue with the test setup or teardown code. For example, if the beforeEach/setup code is taking too long to run, it could cause a timeout in the test. You could try commenting out or simplifying the setup code to see if that resolves the issue.

There might be an issue with the test dependencies or environment setup. For example, if the test is using a particular package or library that is causing issues, you might need to update or remove the package to resolve the issue.

It's also possible that the issue is related to the version of VS Code being used. You mentioned that the test is being run on several different versions of VS Code. It might be worth trying the test on a different version to see if the behavior persists.

Overall, it's difficult to say exactly what is causing the timeout without more information. I recommend starting by adding some debugging statements to see how long the test takes to run, and then looking for any other potential issues in the test environment or setup.

**Reply by kimadeline** [reply]
Hi @Drisate, thank you for your reply!

 > You could try adding some debugging statements to see how long the test takes to run, and compare that to the timeout threshold.

When doing so using Node's [console.time](https://nodejs.org/api/console.html#console_console_timeend_label), the test duration that was reported was the same as Mocha's own reported test length. Besides, the test isn't reported as timing out, but the setup call of the test suite following that one.

> try running the test in a different environment to see if the behavior persists.

The test times out sporadically on a Windows runner when the GitHub Action workflow runs, and 100% of the time on Ubuntu. I cannot reproduce the issue when I run the test locally on my Mac though.

> You could try commenting out or simplifying the setup code to see if that resolves the issue.

That's the "interesting" thing, there's nothing in the setup code being reported as timing out 🤷

**Reply by olguzzar** [reply]
Hi @kimadeline, bumping this since it's been a few months - is this still an issue you need help with?

**Reply by kimadeline** [reply]
Hi @olguzzar, thank you for resurfacing the thread! We couldn't find any useful leads, so we ended up disabling the flaky tests 😔 I'll close the discussion.
